torch/csrc/autograd/engine.cpp

```diff
@@ -498,6 +498,10 @@ void validate_outputs(
     if (c10::typeMetaToScalarType(metadata.options().dtype()) != grads[i].scalar_type()) {
       grads[i] = grads[i].to(c10::typeMetaToScalarType(metadata.options().dtype()));
     }
+    if (output.device() != metadata.device() &&
+        output.dim() == 0) {
+      output = output.to(metadata.device());
```

Comment:
I don't think you want to write into `output` here. You want to replace the Tensor that is contained into `grads` just like it is done above.

Reply:
In this case output is a reference to `grads[i]` and `grads` is a vector that doesn't change throughout this function, so it should make no difference. I have renamed `output` to `grad` and replaced all `grads[i]` with `grad` which IMO makes the code more readable.

Reply:
Ok, I'm not sure about the cpp semantic here. I though `grads[i] = ` would just set the ith entry to be the new Tensor that is given while `grad =` changes the current ith entry inplace.

Reply:
Both statements have the same effect in this case. They will call `at::Tensor::operator=(Tensor&&)` on the underlying object.

I can change it to `grads[i]` if you think that will be less confusing. In that case, I think I should replace all `grad` (formerly `output`) with `grads[i]` for the sake of consistency.

Either way, I will also look at that CI failure. It might be related.

Reply:
@albanD, can you please clarify whether you want me to use `grads[i] = ...` everywhere or whether you are happy with using `grad = ...`, as proposed in the current PR

Sorry for the late response, the CI was a nightmare to figure out (shoutout to Pearu for figuring it out).

Reply:
@albanD could you please clarify what you want me to do here?

Reply:
Hey, the thumbs up above meant that if they do the exact same thing, I'm happy with either :) So whichever feels more natural for you is good.